modules/pwa/class.jetpack-pwa-service-worker.php

```diff
@@ -0,0 +1,212 @@
+<?php
+
+define( 'PWA_SW_QUERY_VAR', 'jetpack_service_worker' );
+define( 'PWA_SW_CONFIG_QUERY_VAR', 'jetpack_service_worker_config' );
+
+class Jetpack_PWA_Service_Worker {
+	private static $__instance = null;
+
+	/**
+	 * Singleton implementation
+	 *
+	 * @return object
+	 */
+	public static function instance() {
+		if ( ! is_a( self::$__instance, 'Jetpack_PWA_Service_Worker' ) ) {
+			self::$__instance = new Jetpack_PWA_Service_Worker();
+		}
+
+		return self::$__instance;
+	}
+
+	/**
+	 * Registers actions
+	 */
+	private function __construct() {
+		add_action( 'jetpack_activate_module_pwa', array( $this, 'module_activate' ) );
+		add_action( 'jetpack_deactivate_module_pwa', array( $this, 'module_deactivate' ) );
+
+		// register WP_Query hooks for manifest and service worker
+		add_filter( 'query_vars', array( $this, 'register_query_vars' ) );
+		add_action( 'template_redirect', array( $this, 'render_service_worker_js' ), 2 );
+		add_action( 'init', array( $this, 'register_assets' ) );
+		add_action( 'init', array( $this, 'register_rewrite_rule' ) );
+		add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_assets' ) );
+
+		// AMP support
+		add_action( 'amp_post_template_head', array( $this, 'render_amp_serviceworker_script' ) );
+		add_action( 'amp_post_template_footer', array( $this, 'render_amp_serviceworker_element' ) );
+	}
+
+	public function register_rewrite_rule() {
+		add_rewrite_rule('service-worker.js$', 'index.php?' . PWA_SW_QUERY_VAR . '=1', 'top');
+	}
+
+	public function module_activate() {
+		$this->register_rewrite_rule();
+		flush_rewrite_rules();
+	}
+
+	public function module_deactivate() {
+		flush_rewrite_rules();
+	}
+
+	public function register_query_vars( $vars ) {
+		$vars[] = PWA_SW_QUERY_VAR;
+		$vars[] = PWA_SW_CONFIG_QUERY_VAR;
+		return $vars;
+	}
+
+	public function render_service_worker_js() {
+		global $wp_query;
+
+		if ( $wp_query->get( PWA_SW_QUERY_VAR ) ) {
+			header( 'Content-Type: application/javascript; charset=utf-8' );
+			// fake localize - service worker is not loaded in page context, so regular localize doesn't work
+			$pwa = Jetpack_PWA::instance();
+			$pwa_vars = array(
+				'admin_url' => admin_url(),
+				'site_url' => site_url(),
+				'site_icon' => $pwa->site_icon_url( 48 ),
+				'sw_config_url' => $this->get_service_worker_config_url(),
+				'images_url' => plugins_url( 'assets/images/', __FILE__ )
+			);
+			echo preg_replace( '/pwa_vars_json/', json_encode( $pwa_vars ), file_get_contents( plugin_dir_path( __FILE__ ) . 'assets/js/service-worker.js' ) );
```

Comment:
Maybe this has been discussed already, but I suggest adding `do_action( 'jetpack_print_service_worker_scripts' )` here so that plugins can add their own scripts, since Jetpack is reserving the `/service-worker.js` endpoint. This action would be analogous to the existing `wp_print_scripts` action.

Otherwise, the only alternative would be for a plugin to add a `template_redirect` action at priority 1 to print something before this `render_service_worker_js` method gets called.

Reply:
Also, instead of doing a `preg_replace()` here which seems somewhat brittle, why not just do:

```php
printf( 'var pwa_vars_json = %s;', wp_json_encode( $pwa_vars ) );
echo file_get_contents( plugin_dir_path( __FILE__ ) . 'assets/js/service-worker.js' );
```

The `pwa_vars_json` variable is already being listed as a global.